File: Graph/go/connected_components_count.go
package graph

// ConnectedComponentsCount returns number of connected components in an undirected graph.
func ConnectedComponentsCount(g Graph) int {
	visited := map[string]struct{}{}
	count := 0
	for node := range g {
		if exploreComponent(g, node, visited) {
			count++
		}
	}
	return count
}

func exploreComponent(g Graph, node string, visited map[string]struct{}) bool {
	if _, seen := visited[node]; seen {
		return false
	}
	visited[node] = struct{}{}

	for _, neighbor := range g[node] {
		exploreComponent(g, neighbor, visited)
	}
	return true
}

File: Graph/ts/connectedComponentsCount.ts
import { Graph } from "./types";

export function connectedComponentsCount(graph: Graph): number {
  const visited = new Set<string>();
  let count = 0;

  for (const node of Object.keys(graph)) {
    if (explore(graph, node, visited)) {
      count += 1;
    }
  }

  return count;
}

function explore(graph: Graph, node: string, visited: Set<string>): boolean {
  if (visited.has(node)) return false;
  visited.add(node);

  for (const neighbor of graph[node] ?? []) {
    explore(graph, neighbor, visited);
  }
  return true;
}
